<!doctype html>
<html>
	<head>
		<meta charset="utf-8">
		<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">
		<meta name="description" content="Girl Develop It Python 2 Course. Presentation theme forked from Hakim El Hattab's reveal.js">
		<meta name="author" content="Girl Develop It">
		<meta name="apple-mobile-web-app-capable" content="yes" />
		<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />

		<!-- Update with GDI Course Name -->
		<title>Python 2: Errors</title>

		<link rel="icon" type="image/x-icon" href="../assets/imgs/favicon.ico" />

		<!-- RevealJS CDN -->
		<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@4.1.0/dist/reset.css">
		<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@4.1.0/dist/reveal.css">
		<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reveal.js@4.1.0/dist/theme/simple.css" id="theme">

		<!-- GDI Custom Theme -->
		<link rel="stylesheet" href="../assets/css/gdi-theme.css">

		<!-- Theme used for syntax highlighted code -->
		<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.1.0/styles/monokai.min.css">
	</head>
	<body>
		<div class="reveal">
			<div class="slides">

				<section class="heading-only" style="padding-top:5%">
                    <h1>Errors</h1>

					<div class="no-print" style="text-align: left; margin-top: 100px; font-size: 70%;">
						Tips for navigating the slides:
						<ul>
							<li>Press O or Escape for overview mode.</li>
							<li>Visit <a href="?print-pdf" target="_blank">this link</a> for a nice printable version</li>
							<li>Press the copy icon on the upper right of code blocks to copy the code</li>
						</ul>
					</div>
				</section>


                <section>
                    <h3>Class outline:</h3>

                    <ul style="font-size:1.0em;">
                        <li>Error types
                        <li>Tracebacks
                        <li>Debugging
                        <li>Exceptions
                    </ul>

                    <!-- Lesson plan
                    12:05 - Files
                    12:15 - Exercise
                    12:30 - CSV Files
                    12:40 - Exercise
                    1:15 - Error types
                    1:25 - Exercise
                    1:40 - Exceptions
                    1:50 - Exercise
                    -->
                </section>

                <section class="heading-only">
                    <h2>Errors</h2>
                </section>

                <section>
                    <h3>Types of errors</h3>

                    <p>These are common to all programming languages:</p>
                    <ul>
                        <li>Logic errors</li>
                        <li>Syntax errors</li>
                        <li>Runtime errors</li>
                    </ul>
                </section>

                <section class="heading-only">
                    <h2>Logic errors</h2>
                </section>

                <section>
                    <h3>Logic errors</h3>

                    <p>A program has a logic error if it does not behave as expected.
                        Typically discovered via failing tests or bug reports from users.
                    </p>

                    <p>Spot the logic error:</p>
                    <pre style="font-size:0.9em; margin-bottom: 40px;"><code data-trim data-noescape class="python">
                    # Sum up the numbers from 1 to 10
                    sum = 0
                    for x in range(1, 10):
                      sum += x
                    </code></pre>

                    <p>To avoid the wrath of angry users, <a target="_blank" href="https://replit.com/@pamelafox/SumTenTested">write tests</a>.</p>
                </section>

                <section>
                    <h3>Exercise</h3>

                    <ul>
                        <li><a target="_blank" href="https://replit.com/team/gdi01/ErrorsFixSumUp">FixerUpper: Sum Up</a></li>
                        <li><a target="_blank" href="https://replit.com/team/gdi01/ErrorsFixCountPositive">FixerUpper: Count Positive</a></li>
                        <li><a target="_blank" href="https://replit.com/team/gdi01/ErrorsFixGetIndex">FixerUpper: Get Index</a></li>
                    </ul>

                    <p>üí¨  If you need any tips or help, just DM on Zoom and say 'I'm stuck on...' or 'Can you give me a tip on...'</p>

                </section>

                <section class="heading-only">
                    <h2>Syntax errors</h2>
                </section>

                <section>
                    <h3>Syntax errors</h3>

                    <p>Each programming language has syntactic rules.
                        If the rules aren't followed, the program cannot be parsed
                        and will not be executed at all.</p>

                    <p>Spot the syntax errors:</p>
                    <pre style="font-size:0.7em;"><code data-trim data-noescape class="python">
                    if x > 5 <span class="fragment"># Missing colon</span>
                       x += 1
                    </code></pre>

                    <pre style="font-size:0.7em; margin-bottom: 40px;"><code data-trim data-noescape class="python">
                    sum = 0
                    for x in range(0, 10):
                       sum + = x <span class="fragment"># No space needed between + and =</span>
                    </code></pre>

                    <p>To fix a syntax error, read the message carefully and go through your code with a critical eye. üëÅ </p>
                </section>

                <section>
                    <h3><code>SyntaxError</code></h3>

                    <p class="smaller"><em>What it technically means:</em>
                        <br>
                        The file you ran isn‚Äôt valid python syntax
                    </p>

                    <p class="smaller"><em>What it practically means:</em>
                        <br>
                        You made a typo
                    </p>

                    <p class="smaller"><em>What you should look for:</em>
                            <ul class="smaller">
                                <li>Extra or missing parenthesis
                                <li>Missing colon at the end of an if, while, def statements, etc.
                                <li>You started writing a statement but forgot to put any clauses inside
                            </ul>
                    </p>

                    <p class="smaller"><em>Examples:</em></p>
                    <pre style="font-size:0.7em;"><code data-trim data-noescape class="python">
                    print("just testing here"))
                    </code></pre>
                    <pre style="font-size:0.7em;"><code data-trim data-noescape class="python">
                    title = 'Hello, ' + name ', how are you?'
                    </code></pre>

                </section>

                <section>
                    <h3><code>IndentationError</code>/<code>TabError</code></h3>

                    <p class="smaller"><em>What it technically means:</em>
                        <br>
                        The file you ran isn't valid Python syntax, due to indentation inconsistency.
                    </p>

                    <p class="smaller"><em>What it sometimes means:</em>
                        <br>
                        You used the wrong text editor (or one with different settings)
                    </p>

                    <p class="smaller"><em>What you should look for:</em>
                        <ul class="smaller">
                            <li>A typo or misaligned block of statements
                            <li>A mix of tabs and spaces
                                <ul>
                                    <li>Open your file in an editor that shows them
                                    <li><code>cat -A filename.py</code> will show them
                                </ul>
                        </ul>
                    </p>

                    <p class="smaller"><em>Example:</em></p>

                    <pre style="font-size:0.7em;"><code data-trim data-noescape class="python">
                    def sum(a, b):
                        total = a + b
                    	return total
                    </code></pre>
                </section>

                <section class="heading-only">
                    <h2>Runtime errors</h2>
                </section>

                <section>
                    <h3>Runtime errors</h3>

                    <p>A runtime error happens <em>while</em> a program is running,
                        often halting the execution of the program.
                     Each programming language defines its own runtime errors.
                    </p>

                    <p>Spot the runtime error:</p>
                    <pre style="font-size:0.7em; margin-bottom: 40px;"><code data-trim data-noescape class="python">
                    def div_numbers(dividend, divisor):
                        return dividend/divisor

                    quot1 = div_numbers(10, 2)
                    quot2 = div_numbers(10, 1)
                    quot3 = div_numbers(10, 0)  <span class="fragment"># Cannot divide by 0!</span>
                    quot4 = div_numbers(10, -1)
                    </code></pre>

                    <p>To prevent runtime errors, code defensively and write tests for all edge cases.</p>
                </section>



                <section>
                    <h3 style="font-size:1.4em;"><code>TypeError:'X' object is not callable</code></h3>

                    <p class="smaller"><em>What it technically means:</em>
                        <br>
                        Objects of type X cannot be treated as functions
                    </p>

                    <p class="smaller"><em>What it practically means:</em>
                        <br>
                        You accidentally called a non-function as if it were a function
                    </p>

                    <p class="smaller"><em>What you should look for:</em>
                        <ul class="smaller">
                            <li>Parentheses after variables that aren't functions
                            <!--
                            <li>Variables that should be functions being assigned to non-functions
                            <li>Local variables that do not contain functions having the same name
                            as functions in the global frame
                            -->
                        </ul>
                    </p>

                    <p class="smaller"><em>Example:</em></p>

                    <pre style="font-size:0.7em;"><code data-trim data-noescape class="python">
                    sum = 2 + 2
                    sum(3, 5)
                    </code></pre>
                </section>


                <section>
                    <h3><code>...NoneType...</code></h3>

                    <p class="smaller"><em>What it technically means:</em>
                        <br>
                        You used None in some operation it wasn't meant for
                    </p>

                    <p class="smaller"><em>What it practically means:</em>
                        <br>
                        You forgot a return statement in a function
                    </p>

                    <p class="smaller"><em>What you should look for:</em>
                        <ul class="smaller">
                            <li>Functions missing return statements
                            <li>Printing instead of returning a value
                        </ul>
                    </p>

                    <p class="smaller"><em>Example:</em></p>

                    <pre style="font-size:0.7em;"><code data-trim data-noescape class="python">
                    def sum(a, b):
                        print(a + b)

                    total = sum( sum(30, 45), sum(10, 15) )
                    </code></pre>

                </section>

                <section>
                    <h3><code>NameError</code></h3>

                    <p class="smaller"><em>What it technically means:</em>
                        <br>
                        Python looked up a name but couldn't find it
                    </p>

                    <p class="smaller"><em>What it practically means:</em>
                        <ul class="smaller">
                            <li>You made a typo
                            <!--<li>You are trying to access variables from the wrong frame-->
                        </ul>
                    </p>

                    <p class="smaller"><em>What you should look for:</em>
                        <ul class="smaller">
                            <li>A typo in the name
                            <!--<li>The variable being defined in a different frame than expected-->
                        </ul>
                    </p>

                    <p class="smaller"><em>Example:</em></p>

                    <pre style="font-size:0.7em;"><code data-trim data-noescape class="python">
                    fav_nut = 'pistachio'
                    best_chip = 'chocolate'
                    trail_mix = Fav_Nut + best__chip
                    </code></pre>

                </section>

                <!--<section>
                    <h3><code>UnboundLocalError</code></h3>

                    <p class="smaller"><em>What it technically means:</em>
                        <br>
                        A variable that's local to a frame was used before it was assigned
                    </p>

                    <p class="smaller"><em>What it practically means:</em>
                        <br>
                        You are trying to both use a variable from a parent frame,
                        and have the same variable be a local variable in the current frame
                    </p>

                    <p class="smaller"><em>What you should look for:</em>
                        <br>
                        Assignments statements after the variable name
                    </p>

                    <p class="smaller"><em>Example:</em></p>

                    <pre style="font-size:0.7em;"><code data-trim data-noescape class="python">
                    </code></pre>

                </section>-->


                <section class="heading-only">
                    <h2>TraceBacks</h2>
                </section>

                <section>
                    <h3>What's a traceback?</h3>

                    <p>When there's a runtime error in your code,
                        you'll see a <strong>traceback</strong> in the console.</p>

                    <pre style="font-size:0.7em; margin-bottom: 40px;"><code data-trim data-noescape class="python">
                        def div_numbers(dividend, divisor):
                            return dividend/divisor

                        quot1 = div_numbers(10, 2)
                        quot2 = div_numbers(10, 1)
                        quot3 = div_numbers(10, 0)
                        quot4 = div_numbers(10, -1)
                    </code></pre>

                    <pre style="font-size:0.7em; margin-bottom: 40px;"><code data-trim data-noescape class="no-highlight" style="color:black;">
                    Traceback (most recent call last):
                        File "main.py", line 14, in &lt;module&gt;
                            quot3 = div_numbers(10, 0)
                        File "main.py", line 10, in div_numbers
                            return dividend/divisor
                    ZeroDivisionError: division by zero
                    </code></pre>
                </section>

                <section>
                    <h3>Parts of a Traceback</h3>

                    <ul>
                        <li><span style="color:red">The error message itself</span>
                        <li><span style="color:green">Lines #s on the way to the error</span>
                        <li><span style="color:blue">What‚Äôs on those lines</span>
                    </ul>

                    <p>The most recent line of code is always last (right before the error message).</p>

                    <pre style="font-size:0.7em; margin-bottom: 40px;"><code data-trim data-noescape class="no-highlight" style="color:black;">
                    Traceback (most recent call last):
                        File "main.py", <span style="color:green">line 14</span>, in &lt;module&gt;
                            <span style="color:blue">quot3 = div_numbers(10, 0)</span>
                        File "main.py", <span style="color:green">line 10</span>, in div_numbers
                            <span style="color:blue">return dividend/divisor</span>
                    <span style="color:red">ZeroDivisionError: division by zero</span>
                    </code></pre>
                </section>


                <section>
                    <h3>Reading a Traceback</h3>

                    <ol>
                        <li>Read the error message (remember what common error messages mean!)
                        <li>Look at <span style="color:blue">each line</span>, bottom to top,
                            and see if you can find the error.
                    </ol>

                    <pre style="font-size:0.7em; margin-bottom: 40px;"><code data-trim data-noescape class="no-highlight" style="color:black;">
                    Traceback (most recent call last):
                        File "main.py", <span style="color:green">line 14</span>, in &lt;module&gt;
                            <span style="color:blue">quot3 = div_numbers(10, 0)</span>
                        File "main.py", <span style="color:green">line 10</span>, in div_numbers
                            <span style="color:blue">return dividend/divisor</span>
                    <span style="color:red">ZeroDivisionError: division by zero</span>
                    </code></pre>
                </section>

                <section>
                    <h3>Exercise</h3>

                    <p>Fix up all these projects:</p>
                    <ul>
                        <li><a target="_blank" href="https://replit.com/team/gdi01/ErrorsBrokenSongify">Broken Songify</a></li>
                        <li><a target="_blank" href="https://replit.com/team/gdi01/ErrorsBrokenFiles">Broken Files</a></li>
                        <li><a target="_blank" href="https://replit.com/team/gdi01/ErrorsBrokenStrings">Broken Strings</a></li>
                        <li><a target="_blank" href="https://replit.com/team/gdi01/ErrorsBrokenCounts">Broken Counts</a></li>
                        <li><a target="_blank" href="https://replit.com/team/gdi01/ErrorsBrokenSums">Broken Sums</a></li>
                    </ul>

                    <p>Plus an <a target="_blank" href="https://replit.com/team/gdi01/ErrorsErrorQuiz">Errors Quiz</a></p>

                    <p>üí¨  If you need any tips or help, just DM on Zoom and say 'I'm stuck on...' or 'Can you give me a tip on...'</p>
                </section>

                <section class="heading-only">
                    <h2>Exceptions</h2>
                </section>

                <section>
                    <h3>Exceptions</h3>

                    <p>Python raises an <strong>exception</strong> whenever a runtime error occurs.
                    An exception is a mechanism in many languages used to declare and respond to "exceptional" conditions.</p>

                    <p>How an exception is reported:</p>
                    <pre style="font-size:0.7em; margin-bottom: 40px;"><code data-trim data-noescape class="python">
                    >>> 10/0
                    Traceback (most recent call last):
                        File "&lt;stdin&gt;", line 1, in <module>
                    ZeroDivisionError: division by zero
                    </code></pre>

                    <p>If an exception is not handled, the program stops executing immediately.</p>
                </section>

                <section>
                    <h3>Types of exceptions</h3>

                    <p>A few exception types and examples of buggy code:</p>
                    <table>
                        <thead>
                            <tr>
                                <th>Exception
                                <th>Example
                        <tbody>
                            <tr>
                                <td><code>OverflowError</code>
                                <td><code>pow(2.12, 1000)</code>
                            <tr>
                                <td><code>TypeError</code>
                                <td><code>'hello'[1] = 'j'</code>
                            <tr>
                                <td><code>IndexError</code>
                                <td><Code>'hello'[7]</code>
                            <tr>
                                <td><code>NameError</code>
                                <td><code>x += 5</code>
                            <tr>
                                <td><code>FileNotFoundError</code>
                                <td><code>open('dsfdfd.txt')</code>
                    </table>

                    <p>See full list in the <a target="_blank" href="https://docs.python.org/3/library/exceptions.html">exceptions docs</a>.</p>

                </section>

                <section>
                    <h3>The try statement</h3>

                    <p>To handle an exception (keep the program running), use a <code>try</code> statement.</p>

                    <pre style="font-size:1.0em; margin-bottom: 40px;"><code data-trim data-noescape class="python">
                    try:
                        &lt;try suite&gt;
                    except &lt;exception class&gt; as &lt;name&gt;:
                        &lt;except suite&gt;
                       ...
                    </code></pre>

                    <p class="smaller">The <code>&lt;try suite&gt;</code> is executed first.
                        If, during the course of executing the <code>&lt;try suite&gt;</code>, an exception is raised that is not handled otherwise, and
                        If the class of the exception inherits from <code>&lt;exception class&gt;</code>, then
                        the <code>&lt;except suite&gt;</code> is executed, with <code>&lt;name&gt;</code> bound to the exception.
                    </p>
                </section>

                <section>
                    <h3>Try statement example</h3>

                    <pre style="font-size:1.0em; margin-bottom: 40px;"><code data-trim data-noescape class="python">
                    try:
                        quot = 10/0
                    except ZeroDivisionError as e:
                        print('handling a', type(e))
                        quot = 0
                    </code></pre>

                    <p class="python-tutor-link"><a target="_blank" href="http://pythontutor.com/visualize.html#code=try%3A%0A%20%20%20%20quot%20%3D%2010/0%0Aexcept%20ZeroDivisionError%20as%20e%3A%0A%20%20%20%20print%28'handling%20a',%20type%28e%29%29%0A%20%20%20%20quot%20%3D%200&cumulative=true&curInstr=0&heapPrimitives=nevernest&mode=display&origin=opt-frontend.js&py=3&rawInputLstJSON=%5B%5D&textReferences=false">Try in PythonTutor</a></p>
                </section>

                <section>
                    <h3>Try inside a function</h3>

                    <pre style="font-size:0.9em; margin-bottom: 40px;"><code data-trim data-noescape class="python">
                    def div_numbers(dividend, divisor):
                        try:
                            quotient = dividend/divisor
                        except ZeroDivisionError:
                            print("Function was called with 0 as divisor")
                            quotient = 0
                        return quotient

                    div_numbers(10, 2)
                    div_numbers(10, 0)
                    div_numbers(10, -1)
                    </code></pre>

                    <p class="python-tutor-link"><a target="_blank" href="http://pythontutor.com/visualize.html#code=def%20div_numbers%28dividend,%20divisor%29%3A%0A%20%20%20%20try%3A%0A%20%20%20%20%20%20%20%20quotient%20%3D%20dividend/divisor%0A%20%20%20%20except%20ZeroDivisionError%3A%0A%20%20%20%20%20%20%20%20print%28%22Function%20was%20called%20with%200%20as%20divisor%22%29%0A%20%20%20%20%20%20%20%20quotient%20%3D%200%0A%20%20%20%20return%20quotient%0A%0Adiv_numbers%2810,%202%29%0Adiv_numbers%2810,%200%29%0Adiv_numbers%2810,%20-1%29&cumulative=true&curInstr=0&heapPrimitives=nevernest&mode=display&origin=opt-frontend.js&py=3&rawInputLstJSON=%5B%5D&textReferences=false">Try in PythonTutor</a></p>
                </section>

                <section>
                    <h3>Handle file exceptions</h3>

                    <pre style="font-size:0.9em; margin-bottom: 40px;"><code data-trim data-noescape class="python">
                    def read_movies(filename):
                        try:
                          with open(filename, newline="") as file:
                            movies = []
                            reader = csv.reader(file)
                            for row in reader:
                              movies.append(row)
                            return movies
                        except Exception as e:
                          print(e)
                          return []

                    read_movies("movies.csv")
                    read_movies("movies2.csv")
                    </code></pre>

                </section>

                <section>
                    <h3>Exercise</h3>

                    <p>We'll start this one together.</p>

                    <p><a target="_blank" href="https://replit.com/team/gdi01/ErrorsExceptionCatcher">Exception catcher</a></p>
                </section>
                <!--
                <section class="heading-only">
                    <h2>Testing</h3>
                </section>

                <section>
                    <h3>Types of tests</h3>
                    <ul>
                        <li><strong>Unit tests</strong>: Does the small units of the system give a correct output for each input?
                        <li><strong>Integration tests</strong>: Do the different parts of the system work together correctly?
                        <li><strong>Smoke tests</strong>: Do the most basic features of the system work?
                        <li><strong>End-to-end tests</strong>: Does the system handle an entire user workflow from start to end?
                        <li><strong>User acceptance tests</strong>: Does the system meet the business requirements?
                        <li><strong>Performance/load tests</strong>: Can the system scale to meet possible demand?
                    </ul>
                </section>

                <section>
                    <h3>Python unit tests</h3>

                    <p class="smaller">It is possible to use <a href="https://docs.python.org/3/library/doctest.html">doctests</a> for unit testing,
                        but it is better to use the <a href="https://docs.python.org/3/library/unittest.html">unit test</a> module
                        to test production code.</p>

                    <p>Make a separate file for testing:</p>
                    <pre style="font-size:0.9em; margin-bottom: 40px;"><code data-trim data-noescape class="python">
                    import unittest

                    from main import div_numbers

                    class TestStringMethods(unittest.TestCase):

                        def test_div_integers(self):
                            self.assertEqual(div_numbers(50, 25), 2)
                            self.assertEqual(div_numbers(-50, 25), -2)

                        def test_div_zero(self):
                            self.assertEqual(div_numbers(50, 0), 0)

                    if __name__ == '__main__':
                            unittest.main()
                    </code></pre>

                </section>
                -->


			</div>
			<footer>
				<!-- Update with Course Name -->
				<div class="copyright">
					<img src="../assets/imgs/gdi-logos/gdi-logo_peach.png" alt="GDI Logo" width="40px" class="noborder"/>
					<p>Python 2</p>
					<a rel="license" href="http://creativecommons.org/licenses/by-nc/3.0/deed.en_US"><img alt="Creative Commons License" style="border-width:0" src="http://i.creativecommons.org/l/by-nc/3.0/80x15.png"/></a>
				</div>
			</footer>
		</div> <!-- End of Reveal wrapper -->
		
		<!-- REVEAL.JS Script Files - Do Not Edit -->
		<script src="https://cdn.jsdelivr.net/npm/reveal.js@4.1.0/dist/reveal.js"></script>
		<script src="https://cdn.jsdelivr.net/npm/reveal.js@4.1.0/plugin/highlight/highlight.js"></script>
		<script src="https://cdn.jsdelivr.net/npm/highlightjs-badge@0.1.9/highlightjs-badge.min.js"></script>
		<script src="https://cdnjs.cloudflare.com/ajax/libs/jsPlumb/2.15.5/js/jsplumb.min.js"></script>
		<script>
			const srcUrlPrefix = "https://cdn.jsdelivr.net/npm/reveal.js@4.1.0/";
			Reveal.initialize({
				hash: true,
				center: true,
				slideNumber: true,
				showNotes: false,
				margin: 0.1,
				preloadIframes: true,
				plugins: [ RevealHighlight ],
				pdfSeparateFragments: true
			});

      // add HighlightJS-badge (options are optional)
			var options = {
					copyIconContent: "üìÑ",
					checkIconContent: "‚úÖ"
			};
			window.highlightJsBadge(options);

			if (window.location.search == "?print-pdf") {
					var uncounted = document.querySelectorAll("[data-visibility='uncounted']");
					uncounted.forEach(node => {
							node.parentNode.classList.add("no-print")
					})
			}
		</script>
	</body>
</html>